### DISCUSSION TITLE
What is the usage of translations in the ScalingMatrix?

### DISCUSSION THREAD

**Discussion by Blatko1** [parasyte/pixels#371] [Q&A] [answered]
I was playing with the project's source code for better understanding, and when I replaced the `tx` and `ty` variables with `0`, the results looked the same. What is the usage of translations in the ScalingMatrix? Calculating the `tx` and `ty` gives some tiny negligible numbers.
I am talking about this specific part of the code:
```rust
let (texture_width, texture_height) = texture_size;
let (screen_width, screen_height) = screen_size;

let width_ratio = (screen_width / texture_width).max(1.0);
let height_ratio = (screen_height / texture_height).max(1.0);

// Get smallest scale size
let scale = width_ratio.clamp(1.0, height_ratio).floor();

let scaled_width = texture_width * scale;
let scaled_height = texture_height * scale;

// Create a transformation matrix
let sw = scaled_width / screen_width;
let sh = scaled_height / screen_height;
let tx = (screen_width / 2.0).fract() / screen_width;
let ty = (screen_height / 2.0).fract() / screen_height;
#[rustfmt::skip]
let transform: [f32; 16] = [
    sw,  0.0, 0.0, 0.0,
    0.0, sh,  0.0, 0.0,
    0.0, 0.0, 1.0, 0.0,
    tx,  ty,  0.0, 1.0,
];
```

**Comment by parasyte** [comment]
The scaling matrix is for adding the border around the image when the surface is not an exact multiple of the image size. A good example is the screenshot used for the `minimal-web` demo:

![minimal-web screenshot](https://github.com/parasyte/pixels/blob/d140eba7c9849372a6673b0d37bfe83948c4e530/img/minimal-web.png?raw=true)

A better example is the Android screenshot:

![minimal-winit-android screen shot](https://github.com/parasyte/pixels/blob/d140eba7c9849372a6673b0d37bfe83948c4e530/img/minimal-winit-android.png?raw=true)

Without the border, the texture would be unnaturally stretched to fit the surface size.

**Reply by Blatko1** [reply]
I understand the usage of `sw` and `sh` variables. As you said, the whole image would be stretched without these variables. I know these variables **scale** or modify the width and height of the *screen triangle* in the shader code, but I am asking for the exact usage of the `tx` and `ty` variables. Their general use would be to move (translate) the image left or right and up or down, but their values are mostly below 0.001, which is negligible. I provided pictures of my findings in the following part.
With transformations removed (`tx` and `ty` variables set to `0`):
![image](https://github.com/parasyte/pixels/assets/51339494/049c953c-c910-40b9-828a-0535f84d6d3b)
With transformations applied (using actual `tx` and `ty`):
![image](https://github.com/parasyte/pixels/assets/51339494/8cb982f5-1764-4d5e-9ed2-b01300d92da7)
As you can see, there is no difference.
I tried to multiply the `tx` and `ty` variables by 100 to see how they impact the result, and all I can notice is that, when resizing the window, the image jiggles left and right by a small amount which is invisible whilst these variables have their typical values. 

Maybe there is no need for the `tx` and `ty` variables?

**Reply by parasyte** [reply, marked as answer]
Specifically, `tx` and `ty` are for [translation](https://en.wikipedia.org/wiki/Translation_(geometry)). If the surface is an odd number of pixels in either direction, these variables will nudge the triangle by half a pixel so that the pixels in the triangle align with pixels on the physical screen.

Without them, you would get an ugly half-pixel offset. See #231.

**Reply by Blatko1** [reply]
Thank you for the explanation, I now see that it fixes a very subtle bug.
